dnf5-plugins/system_upgrade_plugin/system_upgrade.hpp

```diff
@@ -0,0 +1,158 @@
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef DNF5_COMMANDS_SYSTEM_UPGRADE_HPP
+#define DNF5_COMMANDS_SYSTEM_UPGRADE_HPP
+
+#include <dnf5/context.hpp>
+#include <libdnf5/conf/const.hpp>
+#include <libdnf5/conf/option_bool.hpp>
+#include <libdnf5/conf/option_number.hpp>
+#include <sdbus-c++/sdbus-c++.h>
+#include <toml.hpp>
+
+const std::filesystem::path PATH_TO_PLYMOUTH{"/usr/bin/plymouth"};
+
+const std::string STATUS_DOWNLOAD_INCOMPLETE{"download-incomplete"};
+const std::string STATUS_DOWNLOAD_COMPLETE{"download-complete"};
+const std::string STATUS_READY{"ready"};
+const std::string STATUS_UPGRADE_INCOMPLETE{"upgrade-incomplete"};
+
+const int STATE_VERSION = 0;
+const std::string STATE_HEADER{"offline-transaction-state"};
+
+struct OfflineTransactionStateData {
+    int state_version = STATE_VERSION;
+    std::string status = STATUS_DOWNLOAD_INCOMPLETE;
+    std::string cachedir;
+    std::string target_releasever;
+    std::string system_releasever;
+    bool poweroff_after = false;
+    std::vector<std::string> enabled_repos;
+    std::vector<std::string> disabled_repos;
+};
+
+TOML11_DEFINE_CONVERSION_NON_INTRUSIVE(
+    OfflineTransactionStateData,
+    state_version,
+    status,
+    cachedir,
+    target_releasever,
+    system_releasever,
+    poweroff_after,
+    enabled_repos,
+    disabled_repos)
+
+class OfflineTransactionState {
+public:
+    void write();
+    OfflineTransactionState(std::filesystem::path path);
+    OfflineTransactionStateData & get_data() { return data; };
+    const std::exception_ptr & get_read_exception() const { return read_exception; };
+
+private:
+    void read();
+    std::exception_ptr read_exception;
+    std::filesystem::path path;
+    OfflineTransactionStateData data;
+};
+
+const std::filesystem::path DEFAULT_DATADIR{std::filesystem::path(libdnf5::SYSTEM_STATE_DIR) / "system-upgrade"};
+
+namespace dnf5 {
+
+class SystemUpgradeCommand : public Command {
+public:
+    explicit SystemUpgradeCommand(Context & context) : Command(context, "system-upgrade") {}
+    void set_parent_command() override;
+    void set_argument_parser() override;
+    void register_subcommands() override;
+    void pre_configure() override;
+};
+
+class SystemUpgradeSubcommand : public Command {
+public:
+    explicit SystemUpgradeSubcommand(Context & context, const std::string & name);
+    void set_argument_parser() override;
+    void configure() override;
+
+protected:
+    libdnf5::OptionPath * get_cachedir() { return cachedir; };
+    std::filesystem::path get_datadir() { return datadir; };
+    void set_datadir(std::filesystem::path new_datadir) { datadir = std::move(new_datadir); };
```

Comment:
Seems unused.

Reply:
Thanks!

Reply:
Resolved in https://github.com/rpm-software-management/dnf5/pull/1280.